File: Guia3/ej7/BankAccount.java
package Guias.Guia3.ej7;

public abstract class BankAccount {
    private final int id;
    protected double balance;

    public BankAccount(int id){
        this.id = id;
    }

    public void deposit(double amount){
        this.balance += amount;
    }
    public void extract(double amount){
        if(canExtract(amount))
            balance -= amount;
    }
    protected boolean canExtract(double amount){
        return balance >= amount;
    }
    @Override
    public String toString(){
        return "Cuenta %d con saldo %.2f".formatted(id, balance);
    }

}

File: POO-Ruby/Guia8/ej5/bank_account.rb
# frozen_string_literal: true

class BankAccount

  def create(id)
    @id = id
    @balance=0
  end
  def initialize
    raise 'BankAccount is an abstract class'
  end

  def deposit(amount)
    @balance+= amount
  end

  def extract(amount)
    @balance -= amount unless extract? amount
  end

  def extract?(amount)
    raise'Undefined method extract?'
  end

  def to_s
    "Cuenta #{@id} con saldo #{@balance}"
  end

  private :create, :extract?

end
